README
Clinical filtering for trios

Find variants in affected children that might contribute to their disorder. We
load VCF files (either named on the command line, or listed in a PED file) for
members of a family, filter for rare, functionally disruptive variants, and
assess whether each variant might affect the child's disorder. We take into 
account the parents genotypes (if available) and whether the parents are also
affected with a (the?) disorder. For variants in known disease causative genes
we check whether the inheritance patterns matches one expected for the  
inheritance models of the gene.

Standard usage:

python clinical_filter.py \
    --ped temp_name.ped \
    --filter filters.txt \
    --tags tags.txt \
    --syndrome-regions regions_filename.txt \ # optional
    --known-genes known_genes.txt \ # optional (but recommended)
    --known-genes-date 2014-01-01 \ # optional
    --alternate-ids alternate_ids.txt \ # optional
    --output output_name.txt \ # optional
    -export-vcf vcf_filename_or_directory # optional

For running the filtering, the basic command is either with a ped file
specified, i.e.

python clinical_filter.py \
  --ped PED_PATH \
  --filter FILTERS_PATH \
  --tags TAGS_PATH

Or the individual VCF files for a trio can be specified, i.e.

python clinical_filter.py \
  --child CHILD_VCF_PATH \
  --mother MOTHER_VCF_PATH \
  --father FATHER_VCF_PATH \
  --mom-aff MOM_AFFECTED_STATUS (1=unaffected or 2=affected) \
  --dad-aff DAD_AFFECTED_STATUS (1=unaffected or 2=affected) \
  --filter FILTERS_PATH \
  --tags TAGS_PATH

The ped option is the easiest if you have a large number of trios to
process, so you can define all the families and their VCF paths in the ped
file, and run with that.

Other options are:
--syndrome-regions SYNDROMES_PATH # path to file listing DECIPHER regions
--known-genes KNOWN_GENES_PATH # to specify the DDG2P database file
--known-genes-date 2014-01-01 # to specify the version of the known genes file
--alternate-ids ALTERNATE_IDS_PATH # path to file for mapping individuals 
between IDs used in the PED file, to alternate study IDs.
--output OUTPUT_PATH # to specify that you want tab-separated output
written to the given path
--export-vcf OUTPUT_VCF_PATH # to specify you want a filtered VCF, can
give a directory (when analysing multiple individuals), or give a file path

The output options can be omitted, or used together, whichever you need.

CHANGELOG
version 0.0.18 (14 Aug 2014)
    - Changed PP filter from <=0.01 to <0.9

version 0.7 lean (13 Aug 2013)
    - Reduced RAM usage. (< 100 Mb when looking for rare coding variants in exome)
    - This is done by adding function vcf2tsvLean in vcf.py to exclude lines instead of adding them 
      to TSV object

version 0.7 (12 Aug 2013)
    - Fixed a bug from the order of genotypes  'child/mother/father'
    Details: Python was sorting the dictionary keys and  hence father used to appear before mother 
    (F letter before M). The error was in the last two lines of "createMatrix" function.
    
version 0.5 (28 Jan 2013)
    - Change the genotype order from (child/father/mother) to (child / mother / father)
    
version 0.4 (29 Nov 2012)
    - Accept gender information (male or female child) to process X-chr appropriately.
    - Add explict inheritance  model for
        * Autosomal chromosomes and PAR regions on X chr
        * X-chr in males
        * X-chr in females
    - Define PAR regions
    - Previously we had summary table for AR, AD, and AR_CompHet for autosomal chromosomes including
     Y and X. Now, we will add another table for X-chr (either female or male child)

version 0.3 (10 Jan 2012)
    - Move many functions to Parser.py and use it as superclass for Trio class
    - Add optional output for child/mother/father genotype. option -g

version 0.2 (22 Dec 2011)
    - Add support for R output (AR = 1, AR_dbHET = 2 and AD = 3) at first column.
    - Correct compund HET function (remove if all comes from one parent or
      if both has all HETs in child)

version 0.1 (16 Dec 2011)


EVAR_sibs.py
'''
EVAR (Exome Variant Analysis Report) for affected sibs

Given three VCF files of 2 or more affected sibs, return shared variants in at
least two sibs under AR , AR_dbHET and AD.
step 1) create search space by using filters from filters.txt
step 2) create trees based on user defined hierarchy and boundaries in hierarchy.txt
step 3) print report
        a) basic information (data, input files, filters, hierarchies, etc )
        b) Counts of each group under AR, AR_dbHET and AD models
        c) Under each model (AR, AR_dbHET and AD) for each group , print its
           variants (only column defined by the user in output.txt)

Usage :

    python EVAR_sibs_0.1.py \
    -a sib1.vcf,sib2.vcf  \ #(sperated by ',' and no spaces)
    -f filters.txt \
    -o output.txt \
    -r hierarchy.txt \
    -t tags.txt \
    -s R \  # optional to print R-compatible format

Note:
The script accept vcf or vcf.gz files of a single or multiple samples as long
as the all cases or all controls not a mix.

version 0.1 (16 Dec 2011)

sa9@sanger.ac.uk
'''
import numpy as np
import os
import sys
import optparse

import user
import vcf
import parser


class Sibs(parser.Parser):
    def __init__(self,paths,weights_path, filters_path, output_path, tags_path, save):

        self.weights_path = weights_path
        self.filters_path = filters_path
        self.output_path = output_path
        self.paths = paths
        
        self.weights = user.parseWeights(weights_path)
        self.filters = user.parseFilters(filters_path)
        self.orders = user.parseOrders(output_path, weights_path)
        self.GN_tag, self.CQ_tag, self.MAF_tag, self.GT_tag = user.parseTags(tags_path)
        
        self.genes_dict = {}
        self.affectedTSVs = {}
        
        for path in paths:
            sample_name = os.path.basename(path)
            self.affectedTSVs[sample_name] = vcf.vcf2tsv(path)
            
        for sample_name in sorted(self.affectedTSVs.keys()):
            self.parse(self.affectedTSVs[sample_name], sample_name)
        
        self.createMatrix()
        self.get_var_under_models()
        
        if save == None:
            self.printReportInputInfo('\tAffected sibs analysis [EVA Report]')
            self.printStatsTable()
            self.printResults()
        elif save == "R":
            self.save2R()
        elif save == "Excel":
            self.save2Excel()
    
    
    def createMatrix(self):
        print "#Creating genotypes and weights matrices"
        for gene in self.genes_dict.keys():
            sorted_pos = sorted([int(pos) for pos in self.genes_dict[gene]["POSs"].keys()])
            GTs = {}
            WTs = {}
            
            for member in sorted(self.affectedTSVs.keys()):
                if not GTs.has_key(member):
                    GTs[member] = [] # genotypes
                    WTs[member] = [] # weights

            for pos in sorted_pos:
                 for member in GTs.keys():
                    try:
                        record = self.genes_dict[gene]["POSs"][str(pos)][member]["raw"]
                        GT = vcf.translateGT(record[self.GT_tag])
                        WT = self.getWeights(record)
                        
                        GTs[member].append(GT)
                        WTs[member].append(WT)
                    except Exception as e:
                        GTs[member].append(0) # the member does not has variants in this pos i.e. hom ref
                        zeros = []
                        for key in self.weights.keys():
                            zeros.append(0.0)
                        WTs[member].append(zeros)

            self.genes_dict[gene]['GTs'] = np.array([GT_lst[1] for GT_lst in sorted(GTs.items())])
            self.genes_dict[gene]['WTs'] = np.array([WT_lst[1] for WT_lst in sorted(WTs.items())])
    
    def get_var_under_models(self):
        self.AR_results = self.construct_holder()
        self.AR_dbHET_results = self.construct_holder()
        self.AD_results = self.construct_holder()
        
        wt_length = len(self.orders["weights"])
        
        for member in sorted(self.affectedTSVs.keys()):
            header = self.affectedTSVs[member]['header']
            header.extend(['appears in','sample_name'])
            break
        
        for gene in self.genes_dict.keys():
            
            GTs = self.genes_dict[gene]['GTs']
            #print GTs
            # genotypes under compound /double HET needs speical treatment.
            # Copy them for further manipulation 
            GTs_db_HET = GTs
            
            WTs = self.genes_dict[gene]['WTs']
            sorted_pos = sorted([int(pos) for pos in self.genes_dict[gene]["POSs"].keys()])
            cln_idx = GTs.shape[1]
            
            for i in range(0,cln_idx):
                affected = GTs[:,i].tolist()
                Hom_counts = affected.count(2)
                Het_counts = affected.count(1)
                key = tuple(x for x in WTs[0,i]) # only include child weights , row 0                
                
                # for loci with at least two sibs with HOM
                if Hom_counts  >= 2:
                    ids = []            
                    for n, member in enumerate(sorted(self.affectedTSVs.keys())):
                        if affected[n] == 2:
                            record = self.extractUserFields(self.genes_dict[gene]['POSs'][str(sorted_pos[i])][member]['raw'], header)
                            if member not in ids:
                                ids.append(member)
                    
                    record.append(", ".join(ids))
                    record.append(str(Hom_counts))
                    
                    if not self.AR_results.has_key(key):
                        self.AR_results[key] = {gene:{str(sorted_pos[i]):record}}
                    else:
                        if not self.AR_results[key].has_key(gene):
                            self.AR_results[key][gene] = {str(sorted_pos[i]):record}
                        else:
                            self.AR_results[key][gene][str(sorted_pos[i])] = record
                    #print key, "\t".join(record)

                
                # for loci with at least two sibs with HET
                if Het_counts >= 2:
                    ids = []       
                    for n, member in enumerate(sorted(self.affectedTSVs.keys())):
                        if affected[n] == 1:
                            record = self.extractUserFields(self.genes_dict[gene]['POSs'][str(sorted_pos[i])][member]['raw'], header)
                            if member not in ids:
                                ids.append(member)
                            
                    record.append(", ".join(ids))
                    record.append(str(Het_counts))                            
                
                    if not self.AD_results.has_key(key):
                        self.AD_results[key] = {gene:{str(sorted_pos[i]):record}}
                        self.AR_dbHET_results[key] = {gene:{str(sorted_pos[i]):record}}
                    else:
                        if not self.AD_results[key].has_key(gene):
                            self.AD_results[key][gene] = {str(sorted_pos[i]):record}
                            self.AR_dbHET_results[key][gene] = {str(sorted_pos[i]):record}
                        else:
                            self.AD_results[key][gene][str(sorted_pos[i])] = record
                            self.AR_dbHET_results[key][gene][str(sorted_pos[i])] = record
                    #print key, "\t".join(record)

        
        # contintue for double HETs model
        self.compundHetCleaning()
        
    def compundHetCleaning(self):
        '''
        Remove genes with only one HET vairants        
        '''
        for key in self.AR_dbHET_results.keys():
            for gene in self.AR_dbHET_results[key].keys():
                if len(self.AR_dbHET_results[key][gene]) <= 1:
                    del self.AR_dbHET_results[key][gene]



def splitPaths(option, opt, value, parser):
    try:
        setattr(parser.values, option.dest, value.split(','))
    except Exception as e:
        print "For option -a you need at least paths to 2 VCF files sperated by space."
        print e
        sys.exit(0) 

def main():

    parser = optparse.OptionParser()
    parser.add_option('-a', '--affected',  type='string',help='path to affected sibs VCF files)', action="callback", callback=splitPaths)

    
    parser.add_option('-f', '--filter', help='path to filters.txt', default="filters.txt")
    parser.add_option('-r', '--hierarchy', help='path to hierarchy.txt', default="hierarchy.txt")
    parser.add_option('-o', '--output', help='path to output.txt', default="output.txt")
    parser.add_option('-t', '--tags', help='path to tags.txt', default="tags.txt")
    parser.add_option('-s', '--save', help='Excel or R', default=None)

    (opts, args) = parser.parse_args()

    #paths = ["../../EVAR/examples/sibs/sib1_5k.vcf.gz", "../../EVAR/examples/sibs/sib2_5k.vcf.gz","../../EVAR/examples/sibs/sib3_5k.vcf.gz"]
    #opts.affected = '/Users/Macia/Desktop/DropBox/Dropbox/Dropbox/Team29/MyProjects/CHD/Patel_Eammon/data/2011-11-11/Consanguineous/Merged_with_UK10Kmaf/D11_12780.vc_UK10Kmaf.vcf.gz /Users/Macia/Desktop/DropBox/Dropbox/Dropbox/Team29/MyProjects/CHD/Patel_Eammon/data/2011-11-11/Consanguineous/Merged_with_UK10Kmaf/D11_12781.vc_UK10Kmaf.vcf.gz /Users/Macia/Desktop/DropBox/Dropbox/Dropbox/Team29/MyProjects/CHD/Patel_Eammon/data/2011-11-11/Consanguineous/Merged_with_UK10Kmaf/D11_12784.vc_UK10Kmaf.vcf.gz'
    paths = opts.affected.split(",")
    
    
    if None in paths or len(paths) < 2:
        print paths
        print "Please make sure you have at least 2 VCF files. Use option -h to print the help message"
        sys.exit(0)


    filters_path = opts.filter
    weights_path = opts.hierarchy
    output_path = opts.output
    tags_path  = opts.tags

    mySibs = Sibs(paths, weights_path, filters_path, output_path, tags_path, opts.save)

if __name__ == "__main__":
    main()
    

